// @paca-ai/plugin-sdk-react/dist/index.js
import { jsx as u } from "react/jsx-runtime";
import { createContext as c, useContext as l, useMemo as _ } from "react";
import { QueryClient as h, QueryClientProvider as d, useQuery as P } from "@tanstack/react-query";
class $ {
  constructor(e) {
    this.baseUrl = e.baseUrl.replace(/\/$/, ""), this.projectId = e.projectId, this._fetch = e.fetch;
  }
  // ── Core read-only helpers ──────────────────────────────────────────────
  /** List tasks for the current project with optional filters. */
  async listTasks(e = {}) {
    var o, a;
    const t = new URLSearchParams();
    (o = e.status_ids) != null && o.length && t.set("status_ids", e.status_ids.join(",")), (a = e.assignee_ids) != null && a.length && t.set("assignee_ids", e.assignee_ids.join(",")), e.sprint_id && t.set("sprint_id", e.sprint_id), e.parent_task_id && t.set("parent_task_id", e.parent_task_id), e.page && t.set("page", String(e.page)), e.page_size && t.set("page_size", String(e.page_size));
    const s = t.toString(), r = `${this.baseUrl}/projects/${this.projectId}/tasks${s ? `?${s}` : ""}`;
    return (await this._get(r)).tasks;
  }
  /** Get a single task by ID. */
  async getTask(e) {
    return this._get(
      `${this.baseUrl}/projects/${this.projectId}/tasks/${e}`
    );
  }
  /** Get the current project summary. */
  async getProject() {
    return this._get(
      `${this.baseUrl}/projects/${this.projectId}`
    );
  }
  /** List members of the current project. */
  async listMembers() {
    return (await this._get(
      `${this.baseUrl}/projects/${this.projectId}/members`
    )).members;
  }
  // ── Plugin route helpers ────────────────────────────────────────────────
  /**
   * Call a GET route registered by this plugin.
   * The URL is built as `{baseUrl}/plugins/{pluginId}{path}`.
   * For project-scoped routes, include the project prefix in `path`:
   *   `pluginGet(id, \`/projects/${api.projectId}/resource\`)`
   */
  async pluginGet(e, t) {
    return this._get(this._pluginUrl(e, t));
  }
  /**
   * Call a POST route registered by this plugin.
   */
  async pluginPost(e, t, s) {
    return this._request("POST", this._pluginUrl(e, t), s);
  }
  /**
   * Call a PATCH route registered by this plugin.
   */
  async pluginPatch(e, t, s) {
    return this._request("PATCH", this._pluginUrl(e, t), s);
  }
  /**
   * Call a DELETE route registered by this plugin.
   */
  async pluginDelete(e, t) {
    await this._request(
      "DELETE",
      this._pluginUrl(e, t),
      void 0
    );
  }
  // ── Internals ───────────────────────────────────────────────────────────
  _pluginUrl(e, t) {
    const s = t.startsWith("/") ? t : `/${t}`;
    return `${this.baseUrl}/plugins/${e}${s}`;
  }
  async _get(e) {
    return this._request("GET", e, void 0);
  }
  async _request(e, t, s) {
    const r = {
      method: e,
      headers: { "Content-Type": "application/json" }
    };
    s !== void 0 && (r.body = JSON.stringify(s));
    const i = await this._fetch(t, r);
    if (!i.ok) {
      const a = await i.text().catch(() => i.statusText);
      throw new Error(
        `[PluginApiClient] ${e} ${t} → ${i.status}: ${a}`
      );
    }
    return i.status === 204 ? void 0 : (await i.json()).data;
  }
}
const y = {
  toast: () => {
  },
  confirm: () => Promise.resolve(!1),
  navigate: () => {
  }
}, g = c(null);
function j({
  api: n,
  ui: e = y,
  meta: t,
  children: s
}) {
  return /* @__PURE__ */ u(g.Provider, { value: { api: n, ui: e, meta: t }, children: s });
}
function v() {
  const n = l(g);
  if (!n)
    throw new Error("usePlugin must be used inside a <PluginProvider>");
  return n;
}
const p = c(null);
function T({
  children: n,
  queryClient: e
}) {
  const t = _(
    () => new h({
      defaultOptions: {
        queries: { retry: 1, staleTime: 3e4 }
      }
    }),
    []
  ), s = e ?? t;
  return /* @__PURE__ */ u(p.Provider, { value: s, children: /* @__PURE__ */ u(d, { client: s, children: n }) });
}
function f(n, e, t, s) {
  return P({
    ...s,
    queryKey: ["plugin", n, ...e],
    queryFn: t
  });
}
function R() {
  const n = l(p);
  if (!n)
    throw new Error(
      "usePluginQueryClient must be inside a <PluginQueryClientProvider>"
    );
  return n;
}
function U(n) {
  const e = n.toUpperCase();
  return {
    canManageProject: e === "PROJECT_OWNER" || e === "PROJECT_MANAGER",
    canManageMembers: e === "PROJECT_OWNER" || e === "PROJECT_MANAGER",
    canWriteTasks: e === "PROJECT_OWNER" || e === "PROJECT_MANAGER" || e === "PROJECT_MEMBER",
    canReadTasks: !0
  };
}
export {
  y as NoopPluginUI,
  $ as PluginApiClient,
  j as PluginProvider,
  T as PluginQueryClientProvider,
  U as permissionsFromRole,
  v as usePlugin,
  f as usePluginQuery,
  R as usePluginQueryClient
};
